syntax = "proto3";

import "types.proto";

package flame;

/*
  The frontend service of Flame, which is used to communicate with client sdk,
  e.g. create connection/session/task and so on.
 */
service Frontend {
  rpc CreateSession (CreateSessionRequest) returns (Session) {}
  rpc DeleteSession (DeleteSessionRequest) returns (Result) {}

  rpc OpenSession (OpenSessionRequest) returns (Result) {}
  rpc CloseSession (CloseSessionRequest) returns (Result) {}

  rpc GetSession(GetSessionRequest) returns (Session) {}
  rpc ListSession (ListSessionRequest) returns (SessionList) {}

  rpc CreateTask (CreateTaskRequest) returns (Task) {}
  rpc DeleteTask (DeleteTaskRequest) returns (Result) {}

  rpc GetTask (GetTaskRequest) returns (Task) {}
  rpc WatchTask (WatchTaskRequest) returns (stream Task) {}
}

message CreateSessionRequest {
  SessionSpec session = 1;
}

message DeleteSessionRequest {
  string session_id = 1;
}

message OpenSessionRequest {
  string session_id = 1;
}

message CloseSessionRequest {
  string session_id = 1;
}
message GetSessionRequest {
  string session_id = 1;
}

message ListSessionRequest {

}

message CreateTaskRequest {
  TaskSpec task = 1;
}

message DeleteTaskRequest {
  string task_id = 1;
  string session_id = 2;
}

message GetTaskRequest {
  string task_id = 1;
  string session_id = 2;
}

message WatchTaskRequest {
  string task_id = 1;
  string session_id = 2;
}
